M2/Macaulay2/m2/html.m2

```diff
@@ -464,24 +464,47 @@ describeReturnCode = r -> (
 
 runFile := (inf,inputhash,outf,tmpf,desc,pkg,announcechange,usermode,examplefiles) -> ( -- return false if error
      announcechange();
-     stderr << "--making " << desc << " in file " << outf << endl;
+     stderr << "--making " << desc << ( if debugLevel > 0 then " in file " | outf else "" ) << endl;
      if fileExists outf then removeFile outf;
      pkgname := toString pkg;
-     setseed := " --no-randomize";
-     ldpkg := if pkgname != "Macaulay2Doc" then concatenate(" -e 'needsPackage(\"",pkgname,"\", Reload => true, FileName => \"",pkg#"source file","\")'") else "";
-     src := concatenate apply(srcdirs, d -> (" --srcdir ",format d));
-     -- we specify --no-readline because the readline library catches SIGINT:
-     args := "--silent --print-width 77 --stop --int --no-readline" | (if usermode then "" else " -q") | src | setseed | ldpkg;
-     env := "GC_MAXIMUM_HEAP_SIZE=400M ";
-     cmdname := commandLine#0;
-     -- must convert a relative path to an absolute path so we can run the same M2 from another directory while
-     -- running the examples:
-     if match("/",cmdname) then cmdname = toAbsolutePath cmdname;
      tmpf << "-- -*- M2-comint -*- hash: " << inputhash << endl << close; -- must match regular expression below
      rundir := temporaryFileName() | "-rundir/";
-     cmd := ulimit | "cd " | rundir | "; " | env | cmdname | " " | args | " <" | format inf | " >>" | format toAbsolutePath tmpf | " 2>&1";
-     stderr << cmd << endl;
      makeDirectory rundir;
+     -* convert usermode to a number and use bitmask (see below) to decide arguments *-
+     defmode := (1<<20)+(1<<21)+1+(1<<1)+(1<<5)+(1<<6)+(1<<12)+(1<<13)+(1<<14)+(1<<30)+(1<<31)+(1<<32); -* =7519367267 *-
+     usermode = if usermode === true then defmode - 1 else usermode; -* the default without -q *-
+     usermode = if usermode === false or usermode === null then defmode else usermode; -* the default *-
+     usermode = if usermode < 0 then (1<<64) + defmode - usermode else usermode; -* useful for changing the default *-
+     -* returns (" "|arg) if the (m)th bit is set *-
+     readmode := (m, arg) -> if usermode & (1 << m) != 0 then " " | arg else "";
```

Comment:
It does not change the top level behavior, as it is backwards compatible; i.e. true, false, and null values are processed exactly as before.

Reply:
In a future PR I'll move this function out of html.m2, and at that point I might document and export it. For now, it serves a very useful purpose for running tests with arbitrary command line options (e.g. disabling threading, readline, etc.) without having an option table with 30 different values.

Reply:
> It does not change the top level behavior, as it is backwards compatible; i.e. true, false, and null values are processed exactly as before.

It does change the top level behavior, even though it is backward compatible, as you say.

Reply:
> In a future PR I'll move this function out of html.m2, and at that point I might document and export it. For now, it serves a very useful purpose for running tests with arbitrary command line options (e.g. disabling threading, readline, etc.) without having an option table with 30 different values.

We don't have any tests needing that, do we?

Reply:
I do. For ctests, I don't want the output to be captured in a file. This way `ctest -V` can show you the output and highlight where an error occurred based on the given regex.

Reply:
What's wrong with the standard way?: Signal an error and save the output file so you can look at it later.

Reply:
I can think of a few reasons:
- why keep an extra file? I can't find a unit testing framework that logs all outputs, all the time.
- if I'm looking for a regex in the output to interrupt computation, I have to see the output directly (and no, convoluted bash pipes are not the answer)
- what makes `--no-randomize --silent --print-width 77 --stop --int --no-readline` "standard"?
- what if I need to test readline, test without threads/preload/setup/etc, or test the new `--webapp`?

Reply:
All that is beside the point: I wanted to take advantage of ctest features to [debug this](https://github.com/Macaulay2/M2/pull/1096#issuecomment-619660197), so made a backwards compatible change which is intentionally not documented because an unexported `runFile` function in `html.m2` with 9 inputs is definitely not the right place to run tests, so I see it as a temporary patch until there's a proper unit testing framework in, say, `testing.m2`.

Reply:
We need the two build systems to run the same tests the same way, and the change to the UserMode option is not acceptable.  If you need a new way to run tests, we can jointly come up with a better way.  For example, what if we had some test files that are simply shell scripts?  Such a shell script could run M2 with any command line options.

Reply:
> We need the two build systems to run the same tests the same way

Why? The new build system is not designed to simply mimic the old build system. It is designed to make development easier, and part of that is making it easier to run tests. If I want to run only `tests/engine/raw-hilb.m2` 100 times in parallel, I can just do `ctest -R engine/raw-hilb --repeat-until-fail 100 -j4`.

> the change to the UserMode option is not acceptable

What does the change to the UserMode option change for the configure build? Nothing at all. Why are you against a backwards compatible change which enhances flexibility of running tests?

I could copy the `runFile` function and make `runFile2` which does what I need and write a `check2` method to go with it, but that would just be duplicated code. Would that be better?

> If you need a new way to run tests, we can jointly come up with a better way.

I do. We can make a new issue and deal with that in a different PR.

Reply:
> > We need the two build systems to run the same tests the same way
> 
> Why? The new build system is not designed to simply mimic the old build system. It is designed to make development easier, and part of that is making it easier to run tests. If I want to run only `tests/engine/raw-hilb.m2` 100 times in parallel, I can just do `ctest -R engine/raw-hilb --repeat-until-fail 100 -j4`.

What I mean is, that, ignoring bells and whistles, I should be able to issue one command to run all the tests, in either build system, and it should be the same tests, run in the same way.

> 
> > the change to the UserMode option is not acceptable
> 
> What does the change to the UserMode option change for the configure build? Nothing at all. Why are you against a backwards compatible change which enhances flexibility of running tests?

It's an egregious kludge, in that it assigns arbitrary meanings to 18 numbers.  And there's certainly a better way, such as a I used here: https://github.com/Macaulay2/M2/blob/master/M2/Macaulay2/tests/normal/interrupt-handling.m2 , or what I suggested above.  And I can't even find an instance where you're using this new facility.


> 
> I could copy the `runFile` function and make `runFile2` which does what I need and write a `check2` method to go with it, but that would just be duplicated code. Would that be better?

No, that would be even worse, with two new functions with non-intuitive names.

> 
> > If you need a new way to run tests, we can jointly come up with a better way.
> 
> I do. We can make a new issue and deal with that in a different PR.

Good.  Meanwhile, put back the old behavior of UserMode.

Reply:
> I should be able to issue one command to run all the tests, in either build system, and it should be the same tests, run in the same way.

You can run `ninja check-packages` and the exact same thing will be executed; no more and no less. For those in `tests` directory, `ctest -R normal` or `ctest -R engine` run the same tests, no more and no less.

> It's an egregious kludge, in that it assigns arbitrary meanings to 18 numbers.

That's what a bitmask is, and used in a ton of places to make things more readable. Wouldn't you say the entire `html.m2` an "egregious kludge"?

> And I can't even find an instance where you're using this new facility.

It is set [here](https://github.com/mahrud/M2/blob/feature/cmake/M2/Macaulay2/packages/CMakeLists.txt#L40-L41), then incorporated in [`M2_TEST_TEMPLATE`](https://github.com/mahrud/M2/blob/98f5be6f041bfaf488ecd0d68edee192972069f3/M2/Macaulay2/packages/CMakeLists.txt#L70), which is then used [here](https://github.com/mahrud/M2/blob/98f5be6f041bfaf488ecd0d68edee192972069f3/M2/Macaulay2/packages/CMakeLists.txt#L214-L216). It is only used if you want to run individual package tests with `ctest -R check-LocalRings-3` for instance. You can also change it by setting the environment variable `UserMode`, which has been invaluable. I'm happy to explain how it works.

> No, that would be even worse, with two new functions with non-intuitive names.

Glad we agree on something.

> Good. Meanwhile, put back the old behavior of UserMode.

Like I said, the behavior hasn't changed. "UserMode => true", "UserMode => false", and "UserMode => null" do the same exact thing, and nowhere in the code you assign an integer to `UserMode` I don't want to refactor the `m2` directory in this PR.

Reply:
> And there's certainly a better way, such as a I used [here](https://github.com/Macaulay2/M2/blob/master/M2/Macaulay2/tests/normal/interrupt-handling.m2) or what I suggested above.

I'm open to different suggestions that would allow me to modify the arguments on any test, without changing the test or the code. With my suggested method, you can pick and choose the arguments and run any test on any package you want. The suggestions you linked are specific to one test or file.

By the way, I think starting a fresh instance of M2 for every test or example when we have threads is a great waste, but that's for another PR, which we won't get to until we get past this.

Reply:
> > It's an egregious kludge, in that it assigns arbitrary meanings to 18 numbers.
> 
> That's what a bitmask is, and used in a ton of places to make things more readable. 

But in this case, you've made things *less* readable.  

Yes, bitmasks can make things more readable.  Here's how.  Suppose you have a function `f` with 5 Boolean arguments.  Then you might call it with `f(0,0,1,0,1)`.  Less cryptic would be `f(false, false, true, false, true)`.  That's still pretty cryptic because no one can remember the order of the arguments.  In a language with optional named arguments, one might be able to type `f(FOO=>true,BAR=>true)`.  That's an improvement, but it's still verbose.  Introducing a bitmask, one can reduce 5 arguments to 1 and write something like `f(0b10100)` or `f((1<<4)|(1<<2))`.  Not very readable, but one can introduce symbols for the appropriate powers of 2 and write `f(FOO|BAR)` instead.  Now we have something readable.  Lots of functions written in low level languages such as C use this method.

In the current pull request, you are taking a step backward in readability, replacing something already symbolic, such as `--notify`, by a hard-to-remember number, namely `11`.  And then in your cmake files you are forced to type this ugly thing:
```
    set(UserMode "(1<<31)+(1<<32)-(1<<64)" # -(1<<8)" # for --no-threads
      CACHE INTERNAL "bitmask for ctest check arguments")
```

Reply:
> This line in html.m2 is particularly unreadable

I totally agree; however, as you always say, the code explains itself: if you read further, each `readmode(n, "--arg")` checks if the n-th bit is set, and if so adds "--arg", and a list of all arguments and the bit associated to them--which has the same order as in startup.m2--comes immediately after. So for instance `(1<<5)` corresponds to `--no-randomize`.

> one can introduce symbols for the appropriate powers of 2 and write f(FOO|BAR) instead. Now we have something readable. Lots of functions written in low level languages such as C use this method.

I'm happy to make this change. The only reason I didn't do that initially is that, when calling `check` from toplevel, we don't have access to those symbols, but at the very least it'll make the line above more readable.

Reply:
> > Good. Meanwhile, put back the old behavior of UserMode.
> 
> Like I said, the behavior hasn't changed. "UserMode => true", "UserMode => false", and "UserMode => null" do the same exact thing, and nowhere in the code you assign an integer to `UserMode` I don't want to refactor the `m2` directory in this PR.

Here's an example that shows the behavior has changed.  Here is the behavior before:
```
+ M2 --no-readline -q --print-width 189
Macaulay2, version 1.15
with packages: ConwayPolynomials, Elimination, IntegralClosure, InverseSystems, LLLBases, PrimaryDecomposition, ReesAlgebra, TangentCone, Truncations

i1 : check ("FirstPackage", UserMode => 1)
--running test 0 of package FirstPackage on line 54 in file /Applications/Macaulay2-1.15/share/Macaulay2/FirstPackage.m2
--    rerun with: check_0 "FirstPackage"
--making test results in file /var/folders/46/9b86vqxj4hjcngvy7kd7sb140000gn/T/M2-59675-0/0.out
stdio:1:1:(3): error: expected true or false

i2 : 
```
And here is your proposed behavior:
```
+ /Users/dan/src/M2/M2.git/M2/BUILD/dan/builds.tmp/einsteinium-1190/M2 --no-readline --print-width 189 -q
Macaulay2, version 1.15.1.0
with packages: ConwayPolynomials, Elimination, IntegralClosure, InverseSystems, LLLBases, PrimaryDecomposition, ReesAlgebra, TangentCone, Truncations

i1 : check ("FirstPackage", UserMode => 1)
--running test 0 of package FirstPackage on line 54 in file ../../../../Macaulay2/packages/FirstPackage.m2
--    rerun with: check_0 "FirstPackage"
--making test results
Macaulay2, version 1.15.1.0
with packages: ConwayPolynomials, Elimination, IntegralClosure, InverseSystems, LLLBases, PrimaryDecomposition, ReesAlgebra, TangentCone, Truncations

i1 : exit

i2 : exit

Process M2<1> finished
```
Notice that the running M2 process has offered the user a prompt -- that's why it took two uses of `exit` to get out.

That behavior is going to be hard to document!

Reply:
> > one can introduce symbols for the appropriate powers of 2 and write f(FOO|BAR) instead. Now we have something readable. Lots of functions written in low level languages such as C use this method.
> 
> I'm happy to make this change. The only reason I didn't do that initially is that, when calling `check` from toplevel, we don't have access to those symbols, but at the very least it'll make the line above more readable.

No, please don't make that change: it would make things worse.  There's no reason to introduce a synonym for something like `--no-readline`, which is already perfectly readable.  Adding redundant symbols adds an unnecessary burden, and they would only be available in that one cmake file.

Reply:
> I'm open to different suggestions that would allow me to modify the arguments on any test, without changing the test or the code. With my suggested method, you can pick and choose the arguments and run any test on any package you want. 

First tell me why this facility would be useful for our users.  I don't see, for example, why it would be useful to enable readline when stdin comes from a file.  It would make no difference.